homeassistant/components/synology_dsm/coordinator.py

```diff
@@ -0,0 +1,148 @@
+"""synology_dsm coordinators."""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+
+import async_timeout
+from synology_dsm.api.surveillance_station.camera import SynoCamera
+from synology_dsm.exceptions import SynologyDSMAPIErrorException
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_SCAN_INTERVAL
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .common import SynoApi
+from .const import (
+    DEFAULT_SCAN_INTERVAL,
+    SIGNAL_CAMERA_SOURCE_CHANGED,
+    SYNOLOGY_CONNECTION_EXCEPTIONS,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SynologyDSMUpdateCoordinator(DataUpdateCoordinator):
+    """DataUpdateCoordinator base class for synology_dsm."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        api: SynoApi,
+        update_interval: timedelta,
+    ) -> None:
+        """Initialize synology_dsm DataUpdateCoordinator."""
+        self.api = api
+        self.entry = entry
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=f"{entry.title} {self.__class__.__name__}",
+            update_interval=update_interval,
+        )
+
+
+class SynologyDSMSwitchUpdateCoordinator(SynologyDSMUpdateCoordinator):
+    """DataUpdateCoordinator to gather data for a synology_dsm switch devices."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        api: SynoApi,
+    ) -> None:
+        """Initialize DataUpdateCoordinator for switch devices."""
+        super().__init__(hass, entry, api, timedelta(seconds=30))
+        self.version: str | None = None
+
+    async def async_setup(self) -> None:
+        """Set up the coordinator initial data."""
+        info = await self.hass.async_add_executor_job(
+            self.api.dsm.surveillance_station.get_info
+        )
+        self.version = info["data"]["CMSMinVersion"]
+
+    async def _async_update_data(self) -> dict[str, dict[str, SynoCamera]] | None:
```

Comment:
When does it return `None`?

Reply:
The dict return value is also typed wrong.

Reply:
Looks like I copied it from the wrong function 👎